hsmtools/tapetohsm.man
.\" @(#)$RCSfile: tapetohsm.man,v $ $Revision: 1.2 $ $Date: 2001/03/28 05:53:15 $ CERN IT-PDP/DM Jean-Philippe Baud
.\" Copyright (C) 2001 by CERN/IT/PDP/DM
.\" All rights reserved
.\"
.TH TAPETOHSM 1 "$Date: 2001/03/28 05:53:15 $" CASTOR "vmgr Administrator Commands"
.SH NAME
tapetohsm \- catalog in the CASTOR Name Server files from a non HSM tape
.SH SYNOPSIS
.B tapetohsm
[
.BI -d " density"
]
.BI -g " dgn"
[
.BI -i
] [
.BI -l " lbltype"
] [
.BI -M " hsm_path_name"
] [
.BI -P " poolname"
] [
.BI -q " file_sequence_number"
] [
.BI -s " size"
]
.BI -V " vid"
[
.BI -v " vsn"
] [
.BI --ml " media_letter"
]
.BI --mo " model"
[
.BI --po " poolname"
] [
.BI --verbose
]
.SH DESCRIPTION
.B tapetohsm
catalogs in the CASTOR Name Server files from a non HSM tape.
.TP
.I density
The supported alphanumeric densities are:
.br
.B 10G
or
.B 10GC
for IBM 3590 cartridges,
.BR 10G ,
.BR 10GC ,
.BR 25G ,
.BR 25GC ,
.B 50G
or
.B 50GC
for Redwood cartridges,
.B 20G
or
.B 20GC
for STK 9840 cassettes.
.TP
.I dgn
The device group name must be a member of an installation defined set of
"device groups".
.TP
.I lbltype
may be
.BR al ,
.B nl
or
.BR sl .
Default is
.BR al .
.TP
.I vid
is the volume visual identifier.
It must be at most CA_MAXVIDLEN characters long.
.TP
.I vsn
specifies the magnetically recorded volume serial number of the tape.
It must be at most six characters long.
.TP
.I media_letter
is the media identification letter. For example "A", "B" or "C" for SD3 (Redwood),
"J" for IBM 3590, "R" for STK 9840.
.TP
.I model
is the model of cartridge.
It must be at most CA_MAXMODELLEN characters long.
.TP
.I poolname
is the tape pool name. It must be at most CA_MAXPOOLNAMELEN characters long.
.LP
Only
.IR vid ,
.I dgn
and
.I model
are mandatory. The other tape fields can be set later.
The file segment attributes can be given on the command line (one file only)
or in a directive file.
.SH EXAMPLE
.nf
.ft CW
tapetohsm -V RT0001 -d 20GC -l al -g 9840R3 --model 9840 --verbose \\
	-M /castor/cern.ch/user/b/baud/RT0001.dat -q 1 -s 111111

tapetohsm -V RT0001 -d 20GC -l al -g 9840R3 --model 9840 --verbose \\
	-i < tapetohsm.inp
.ft
.fi
where tapetohsm.inp contains:
.nf
.ft CW
-q 1 -s 1111111 -M /castor/cern.ch/user/b/baud/RT0001/file1
-q 2 -s 2222222 -M /castor/cern.ch/user/b/baud/RT0001/file2
-q 3 -s 3333333 -M /castor/cern.ch/user/b/baud/RT0001/file3
.ft
.fi
.SH EXIT STATUS
This program returns 0 if the operation was successful or >0 if the operation
failed.
.SH SEE ALSO
.BR Castor_limits(4) ,
.BR Cns_creat(3) ,
.BR Cns_setfsize(3) ,
.BR Cns_setsegattrs(3) ,
.BR vmgr_updatetape(3) ,
.B vmgrentertape(1)


CHANGELOG.md
## 5.11.23.0-1 (2026-09-25)

### New Features

- [frontend] Add user configuration improvements for cta-admin-grpc (cta/CTA!1635)

## 5.11.22.0-1 (2026-09-04)

### New Features

- [frontend] Implement JWT blacklisting on the CTA Frontend (cta/CTA!1548)

- [runtime] Add additional validation logic to application config (cta/CTA!1567)


### Bug Fixes

- [scheduler] Fix issues after pg1 release rest: repack admin commands, archive...

- [Tools] Fix 'priority' column in 'cta-admin dr ls' command (cta/CTA!1546)

- [runtime] Make signals, telemetry and health handling more robust (cta/CTA!1554)

- [runtime] Add bounds checking and better error diagnostics to TOML loading (cta/CTA!1559)

## 5.11.21.0-1 (2026-08-25)

### New Features

- [scheduler] Add scheduler workflow metrics to OpenTelementry (cta/CTA!1387)


### Changes

- [Tools] Move cta-admin af ch command to tf ch (cta/CTA!1525)
- [Tools] Replace cta-objectstore-tools.conf with cta-scheduler.conf (cta/CTA!1540)
- [maintd] Change scheduler.objectstore_backend_path to scheduler.config_file (cta/CTA!1518)


### Bug Fixes

- [frontend] Fix logic bug in frontend authentication and SEGFAULT in cta-admin (cta/CTA!1558)
- [Tools] Fix cta-verify-file (cta/CTA!1545)
- [Tools] Add `--storage-class` option to `cta-send-event` (cta/CTA!1542)


### Performance Improvements

- [taped] Skip block reposition request if drive is already positioned correctly (cta/CTA!1527)

## 5.11.20.0-1

### New Features

- [Tools] Add optional storage class statistics, per tape, to cta-admin tool (cta/CTA!1410)
- [Misc] Add event_name to important CTA log messages (cta/CTA!1381)

### Changes

- [Tools] Change short option for cta-admin af ch --idfile from -I to -F (cta/CTA!1453)
- [taped] Taped no longer sets its own user, group and SYS_RAWIO capability (cta/CTA!1426)
- [Misc] Improve .service and .logrotate configuration (cta/CTA!1524)
- [Misc] Modernise RPM spec metadata and file handling (cta/CTA!1507)
- [Misc] Move systemd units to `/usr/lib/systemd/system` (cta/CTA!1497)
- [Misc] Change systemd to weak RPM dependency of CTA (cta/CTA!1509)
- [Misc] Change logrotate to weak RPM dependency of CTA (cta/CTA!1505)
- [Misc] Remove RPM changelog; package CHANGELOG.md as doc (cta/CTA!1508)
- [Misc] Remove noreplace from example configuration files (cta/CTA!1506)
- [taped] Drive process crashing produces CRIT message instead of INFO (cta/CTA!1499)
- [taped] cta-eosdf no longer sources sysconfig (cta/CTA!1455)

### Bug Fixes

- [frontend] Forbid reclaiming tapes with recently deleted files (cta/CTA!1418)
- [frontend] PREPARE gRPC request shouldn't fail without a storage class (cta/CTA!1471)
- [taped] Fix crash due to accidental pointer arithmetic in exception msg (cta/CTA!1502)
- [taped] Fix opentelemetry shutdown race condition with objectstore (cta/CTA!1500)
- [scheduler]: Fix archive unmount decision logic for postgres scheduler (cta/CTA!1450)
- [frontend] Fix frontend catalogue cache for mount rules (cta/CTA!1490)
- [CI] Ensure CTA images trust CERN Root CA (cta/CTA!1535)
- [Misc] Ensure CMake correctly installs executables as PROGRAMS (cta/CTA!1523)

### Security

- [CI] Generate SBOM for each Docker image (cta/CTA!1532)

### Other

- [CI] Add system tests to cover staging tokens on EOS-CTA (cta/CTA!1493)

## 5.11.19.0-1

### Changes

- [frontend] Split Admin/WFE frontends, add mTLS (cta/CTA!1367)
- [catalogue] Forbid adding or changing encryption key of non-empty tapes (cta/CTA!1414)

### Deprecations

- [frontend] Deprecate storage class field in EOS-CTA xattrs (cta/CTA!1424)

### Other

- [CI] Add Enstore and EnstoreLarge tape tests (cta/CTA!1412)

## 5.11.18.0-2

### Bug Fixes

- [Misc] Update required runtime version of opentelemetry-cpp to v1.26.0 (cta/CTA!1370)

## 5.11.18.0-1

### New Features

- [maintd] Add catalogue and telemetry conf files to runtime directory (cta/CTA!1303)
- [Misc] Package basic JSON logging schema (cta/CTA!1304)
- [Misc] Add source location and CTA version to log lines (cta/CTA!1348)

### Changes

- [frontend] Allow admin commands or physics workflows to be disabled on CTA frontend (cta/CTA!1336)
- [Misc] Replace reflect-cpp library with custom implementation (cta/CTA!1291)
- [Misc] Change Catalogue backend install behaviour (cta/CTA!1295)

### Bug Fixes

- [catalogue] Improve CTA resilience to catalogue DB outages during data fetching (cta/CTA!1334)
- [taped] Improve readExactBlock error message contents (cta/CTA!1311)
- [taped] Enabling reading of tapes with misformed OSM tape format (cta/CTA!1343)
- [frontend] Prevent 'cta-admin dr rm' command from crashing the cta-frontend (cta/CTA!1305)
- [frontend] Remove race conditions in NegotiationRequestHandler (cta/CTA!1310)
- [frontend] Fix broken cta-admin fr ls command (cta/CTA!1340)
- [Misc] Fix compile issues when building CTA in Release mode (cta/CTA!1207)

### Performance Improvements

- [CI] Use python xrootd persistent clients for archive instead of xrdcp --parallel (cta/CTA!1274)

### Other

- [Misc] Update opentelemetry-cpp to v1.26.0 (cta/CTA!1321)


## 5.11.17.0-1

### New Features

- [maintd] Add CLI option to use runtime directory for app metadata (cta/CTA!1266)

### Changes

- [maintd] Update maintd to use TOML configuration (cta/CTA!1246)
- [maintd] Use declarative config to initialise OpenTelemetry SDK (cta/CTA!1248)
- [maintd] Add use_unix_domain_socket configuration entry for health server (cta/CTA!1282)
- [maintd] Remove daemon_user and daemon_group settings (cta/CTA!1252)

### Bug Fixes

- [frontend] Fix crash in 'cta-admin dr ls' command when using multiple...
- [frontend] Add error message when trying to change media type of non-empty tape (cta/CTA!1253)

### Removals

- [frontend] Stop using TAPE_POOL.SUPPLY column (cta/CTA!1245)

### Performance Improvements

- [taped] Optimise DriveState queries (cta/CTA!1198)

### Other

- [Misc] Rename *.conf.example files to *.example.conf (cta/CTA!1279)

## 5.11.16.1-1

### Bug Fixes


hsmtools/tapetohsm.c
/*
 * Copyright (C) 2001-2002 by CERN/IT/PDP/DM
 * All rights reserved
 */

/*	tapetohsm - catalog in the CASTOR Name Server files from a non HSM tape */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#if defined(_WIN32)
#include <winsock2.h>
#endif
#include "Cgetopt.h"
#include "Cns_api.h"
#include "serrno.h"
#include "u64subr.h"
#include "vmgr_api.h"

u_signed64 BytesWritten;
char *hsm_path;
int iflag;
int maxfseq;
struct Cns_segattrs segattrs;
struct vmgr_tape_info tape;
int verbose;

int chkopt(nargs, argv, cmdline)
int nargs;
char **argv;
int cmdline;
{
	int c;
	char *dp;
	int errflg = 0;
	static struct Coptions longopts[] = {
		{"library", REQUIRED_ARGUMENT, 0, OPT_LIBRARY_NAME},
		{"media_letter", REQUIRED_ARGUMENT, 0, OPT_MEDIA_LETTER},
		{"ml", REQUIRED_ARGUMENT, 0, OPT_MEDIA_LETTER},
		{"model", REQUIRED_ARGUMENT, 0, OPT_MODEL},
		{"pool", REQUIRED_ARGUMENT, 0, 'P'},
		{"verbose", NO_ARGUMENT, &verbose, 1},
		{0, 0, 0, 0}
	};

	Copterr = 1;
	Coptind = 1;
	while ((c = Cgetopt_long (nargs, argv, "d:il:M:P:q:s:V:v:", longopts, NULL)) != EOF) {
		switch (c) {
		case OPT_LIBRARY_NAME:
			if (strlen (Coptarg) > CA_MAXTAPELIBLEN) {
				fprintf (stderr,
				    "invalid library %s\n", Coptarg);
				errflg++;
			} else
				strcpy (tape.library, Coptarg);
			break;
		case OPT_MEDIA_LETTER:
			if (strlen (Coptarg) > CA_MAXMLLEN) {
				fprintf (stderr, "%s: %s\n", Coptarg,
				    "media letter too long");
				errflg++;
			} else
				strcpy (tape.media_letter, Coptarg);
			break;
		case OPT_MODEL:
			if (strlen (Coptarg) > CA_MAXMODELLEN) {
				fprintf (stderr, "%s: %s\n", Coptarg,
				    "model name too long");
				errflg++;
			} else
				strcpy (tape.model, Coptarg);
			break;
		case 'd':
			if (strlen (Coptarg) > CA_MAXDENLEN) {
				fprintf (stderr,
				    "invalid density %s\n", Coptarg);
				errflg++;
			} else
				strcpy (tape.density, Coptarg);
			break;
		case 'i':
			if (! cmdline) {
				fprintf (stderr,
				    "option -i is only valid on the command line\n");
				errflg++;
			} else
				iflag++;
			break;
		case 'l':
			if (strlen (Coptarg) > CA_MAXDENLEN) {
				fprintf (stderr,
				    "invalid lbltype %s\n", Coptarg);
				errflg++;
			} else
				strcpy (tape.lbltype, Coptarg);
			break;
		case 'M':
			if (strlen (Coptarg) > CA_MAXPATHLEN) {
				fprintf (stderr, "%s: %s\n", Coptarg,
				    sstrerror(SENAMETOOLONG));
				errflg++;
			} else
				hsm_path = Coptarg;
			break;
		case 'P':
			if (strlen (Coptarg) > CA_MAXPOOLNAMELEN) {
				fprintf (stderr, "%s: %s\n", Coptarg,
				    "pool name too long");
				errflg++;
			} else
				strcpy (tape.poolname, Coptarg);
			break;
		case 'q':
			if ((segattrs.fseq =
			    strtol (Coptarg, &dp, 10)) <= 0 || *dp != '\0') {
				fprintf (stderr,
				    "invalid file sequence number %s\n", Coptarg);
				errflg++;
			}
			break;
		case 's':
			segattrs.segsize = strtou64 (Coptarg);
			if (segattrs.segsize == 0) {
				fprintf (stderr,
				    "invalid size %s\n", Coptarg);
				errflg++;
			}
			break;
		case 'V':
			if (strlen (Coptarg) > CA_MAXVIDLEN) {
				fprintf (stderr,
				    "invalid vid %s\n", Coptarg);
				errflg++;
			} else
				strcpy (tape.vid, Coptarg);
				strcpy (segattrs.vid, Coptarg);
			break;
		case 'v':
			if (strlen (Coptarg) > CA_MAXVSNLEN) {
				fprintf (stderr,
				    "invalid vsn %s\n", Coptarg);
				errflg++;
			} else
				strcpy (tape.vsn, Coptarg);
			break;
		case '?':
			errflg++;
			break;
		default:
			break;
		}
	}
	return (errflg ? USERR : 0);
}

int dir2argv(inpdir, argvp)
char *inpdir;
char ***argvp;
{
	char **argv;
	int c;
	int nargs;
	char *p;
	int parm;

	nargs = 1;
	p = inpdir;
	parm = 0;
	while ((c = *p++)) {
		if (c == ' ' || c == '\t') {
			parm = 0;
		} else if (!parm) {
				parm++;
				nargs++;
		}
	}
	argv = (char **) malloc ((nargs + 1) * sizeof(char *));
	argv[0] = "tapetohsm";
	nargs = 1;
	p = inpdir;
	parm = 0;
	while ((c = *p++)) {
		if (c == ' ' || c == '\t') {
			if (parm) {
				parm = 0;
				*(p - 1) = '\0';
			}
		} else if (!parm) {
			parm++;
			argv[nargs++] = p - 1;
		}
	}
	argv[nargs] = NULL;
	*argvp = argv;
	return (nargs);
}

int procfile()
{
	int errflg = 0;
	struct Cns_fileid file_uniqueid;
	char tmpbuf[21];
	char tmpbuf2[21];

	if (! hsm_path) {
		fprintf (stderr, "hsm_path is mandatory\n");
		return (USERR);
	}

	if (segattrs.segsize == 0) {
		fprintf (stderr, "size is mandatory\n");
		errflg++;
	}
	if (segattrs.fseq == 0)
		segattrs.fseq = 1;

	/* create the HSM file entry in the Name Server */

	if (verbose)
		printf ("Cns_creatx (%s)\n", hsm_path);
	if (Cns_creatx (hsm_path, 0666, &file_uniqueid) < 0) {
		fprintf (stderr, "%s: %s\n", hsm_path, sstrerror(serrno));
		return (USERR);
	}

	/* set filesize */

	if (verbose)
		printf ("Cns_setfsize (NULL, %s, %s)\n",
		    u64tostr (file_uniqueid.fileid, tmpbuf, 0),
		    u64tostr (segattrs.segsize, tmpbuf2, 0));
	if (Cns_setfsize (NULL, &file_uniqueid, segattrs.segsize) < 0) {
		fprintf (stderr, "%s: %s\n", hsm_path, sstrerror(serrno));
		return (USERR);
	}

	/* set segment attributes */

	if (verbose)
		printf ("Cns_setsegattrs (NULL, %s, 1, %s %s %d)\n",
		    u64tostr (file_uniqueid.fileid, tmpbuf, 0),
		    u64tostr (segattrs.segsize, tmpbuf2, 0),
		    segattrs.vid, segattrs.fseq);
	if (Cns_setsegattrs (NULL, &file_uniqueid, 1, &segattrs) < 0) {
		fprintf (stderr, "%s: %s\n", hsm_path, sstrerror(serrno));
		return (USERR);
	}
	if (segattrs.fseq > maxfseq)
		maxfseq = segattrs.fseq;
	BytesWritten += segattrs.segsize;
	hsm_path = NULL;
	segattrs.fseq = 0;
	segattrs.segsize = 0;
	return (0);
}

int procinpdir(s)
FILE *s;
{
	char **argv_d;
	char buf[256];
	int errflg = 0;
	int n;
	int nargs;
	char *p;

	/* check and store the directives in an array */

	while (fgets (buf, sizeof(buf), s) != NULL) {
		if (buf[strlen (buf)-1] != '\n') {
			fprintf (stderr, "directive line too long\n");
			return (USERR);
		} else
			buf[strlen (buf)-1] = '\0';
		if ((n = strspn (buf, " \t")) == strlen (buf)) continue; /* blank line */
		p = buf + n;
		if (*p == '#') continue;	/* comment line */
		nargs = dir2argv (p, &argv_d);
		if (chkopt (nargs, argv_d, 0)) {
			errflg++;
			continue;
		}
		if (procfile ())
			errflg++;
	}
	return (errflg ? USERR : 0);
}

void usage(cmd)
char *cmd;
{
	fprintf (stderr, "usage: %s ", cmd);
	fprintf (stderr, "%s%s%s",
	    "-d density [-i] [-l lbltype] [-M hsm_path_name]\n",
	    "\t[-P pool_name] [-q file_sequence_number] [-s size] -V vid [-v vsn]\n",
	    "\t--li library_name [--ml media_letter] --mo model [--verbose]\n");
}

int main(argc, argv)
int argc;
char **argv;
{
	int c;
	int errflg = 0;
	char tmpbuf[21];
#if defined(_WIN32)
	WSADATA wsadata;
#endif

	segattrs.fsec = 1;
	segattrs.s_status = '-';

	if ((c = chkopt (argc, argv, &segattrs, 1))) {
		usage (argv[0]);
		goto prt_final_status;
	}
	if (*tape.vid == '\0' || *tape.model == '\0' || *tape.library == '\0' ||
	    *tape.density == '\0') {
		c = USERR;
		usage (argv[0]);
		goto prt_final_status;
	}

#if defined(_WIN32)
	if (WSAStartup (MAKEWORD (2, 0), &wsadata)) {
		fprintf (stderr, "WSAStartup unsuccessful\n");
		c = SYERR;
		goto prt_final_status;
	}
#endif
#if TMS
	if (tmscheck (tape.vid, tape.vsn, tape.model, tape.library,
	    tape.density, tape.lbltype)) {
		c = USERR;
		goto prt_final_status;
	}
#endif

	/* enter the tape in the Volume Manager */

	if (verbose)
		printf ("vmgr_entertape (%s, %s, %s, %s, %s, %s, %s, %s, %s, %d, %s, %x)\n",
		    tape.vid, tape.vsn, tape.library, tape.density,
		    tape.lbltype, tape.model, tape.media_letter, "", "", 1,
		    tape.poolname, TAPE_RDONLY);
	if (vmgr_entertape (tape.vid, tape.vsn, tape.library, tape.density,
	    tape.lbltype, tape.model, tape.media_letter, NULL, NULL, 1,
	    tape.poolname, TAPE_RDONLY) < 0) {
		fprintf (stderr, "vmgrentertape %s: %s\n", tape.vid, sstrerror(serrno));
		c = USERR;
		goto prt_final_status;
	}

	if (! iflag)		/* all parameters on command line */
		c = procfile();
	else
		c = procinpdir (stdin);
	if (c)
		goto prt_final_status;

	if (verbose)
		printf ("vmgr_updatetape (%s, 0, %s, 0, %d, 0)\n",
		    tape.vid, u64tostr (BytesWritten, tmpbuf, 0), maxfseq);
	if (vmgr_updatetape (tape.vid, 0, BytesWritten, 0, maxfseq, 0) < 0) {
		fprintf (stderr, "%s: %s\n", tape.vid, sstrerror(serrno));
		errflg++;
	}
	c = errflg ? USERR : 0;
prt_final_status:
#if defined(_WIN32)
	WSACleanup();
#endif
	exit (c);
}

#if TMS
int tmscheck(vid, vsn, model, library, den, lbl)
char *vid;
char *vsn;
char *model;
char *library;
char *den;
char *lbl;
{
	int c, j;
	int errflg = 0;
	char *p;
	int repsize;
	int reqlen;
	char tmrepbuf[132];
	static char tmsden[6] = "     ";
	static char tmslbl[3] = "  ";
	static char tmslibrary[9] = "        ";
	static char tmsmodel[7] = "      ";
	char tmsreq[80];
	static char tmsvsn[7] = "      ";

	sprintf (tmsreq, "QUERY VID %s", vid);
	reqlen = strlen (tmsreq);
	while (1) {
		repsize = sizeof(tmrepbuf);
		c = sysreq ("TMS", tmsreq, &reqlen, tmrepbuf, &repsize);
		switch (c) {
		case 0:
			break;
		case 8:
		case 100:
			fprintf (stderr, "%s\n", tmrepbuf);
			errflg++;
			break;
		default:
			sleep (60);
			continue;
		}
		break;
	}
	if (errflg) return (USERR);
	strncpy (tmsvsn, tmrepbuf, 6);
	for  (j = 0; tmsvsn[j]; j++)
		if (tmsvsn[j] == ' ') break;
	tmsvsn[j] = '\0';
	if (*vsn) {
		if (strcmp (vsn, tmsvsn)) {
			fprintf (stderr,
			    "parameter inconsistency with TMS for vid %s: %s<->%s\n",
			    vid, vsn, tmsvsn);
			errflg++;
		}
	}

	strncpy (tmsmodel, tmrepbuf+ 16, 6);
	for  (j = 0; tmsmodel[j]; j++)
		if (tmsmodel[j] == ' ') break;
	tmsmodel[j] = '\0';
	if (*model) {
		if (strcmp (model, tmsmodel) != 0) {
			fprintf (stderr,
			    "parameter inconsistency with TMS for vid %s: %s<->%s\n",
			    vid, model, tmsmodel);
			errflg++;
		}
	}

	strncpy (tmslibrary, tmrepbuf+ 36, 8);
	for  (j = 0; tmslibrary[j]; j++)
		if (tmslibrary[j] == ' ') break;
	tmslibrary[j] = '\0';
	if (*library) {
		if (strcmp (library, tmslibrary) != 0) {
			fprintf (stderr,
			    "parameter inconsistency with TMS for vid %s: %s<->%s\n",
			    vid, library, tmslibrary);
			errflg++;
		}
	}

	p = tmrepbuf + 23;
	while (*p == ' ') p++;
	j = tmrepbuf + 31 - p;
	strncpy (tmsden, p, j);
	tmsden[j] = '\0';
	if (*den) {
		if (strcmp (den, tmsden) != 0) {
			fprintf (stderr,
			    "parameter inconsistency with TMS for vid %s: %s<->%s\n",
			    vid, den, tmsden);
			errflg++;
		}
	}

	tmslbl[0] = tmrepbuf[32] - 'A' + 'a';
	tmslbl[1] = tmrepbuf[33] - 'A' + 'a';
	if (*lbl) {
		if (strcmp (lbl, tmslbl)) {
			fprintf (stderr,
			    "parameter inconsistency with TMS for vid %s: %s<->%s\n",
			    vid, lbl, tmslbl);
			errflg++;
		}
	}
	return (errflg);
}
#endif
